File: core/renderer.py
import os
import subprocess
from pathlib import Path

from core.downloader import VIDEOS_DIR, CAPTIONS_DIR, get_ffmpeg_path, get_vgmstream_path

def get_filename_no_ext(path):
    return os.path.splitext(os.path.basename(path))[0]

def call_vgmstream(infile, outfile, log_callback=None):
    vgmstream_exe = get_vgmstream_path()
    cmd = [vgmstream_exe, "-o", outfile, infile]
    
    if log_callback:
        process = subprocess.Popen(cmd, stdout=subprocess.PIPE, stderr=subprocess.STDOUT, text=True, bufsize=1, universal_newlines=True)
        for line in iter(process.stdout.readline, ''):
            if line: log_callback("vgmstream: " + line.strip())
        process.wait()
    else:
        subprocess.call(cmd, stdout=subprocess.DEVNULL, stderr=subprocess.DEVNULL)

import concurrent.futures
from threading import Lock

def render_single_video(i, video, total, subtitle_mode, log_callback, progress_callback):
    # 1. Generate soundtracks
    new_sounds = []
    for sound in video["Sound"]:
        wav_name = get_filename_no_ext(sound) + ".wav"
        wav_path = VIDEOS_DIR / wav_name
        
        if not wav_path.exists():
            call_vgmstream(sound, str(wav_path), None) # Mute vgmstream logs in threads to avoid clutter
            
        if wav_path.exists():
            new_sounds.append(str(wav_path))
        else:
            if log_callback: log_callback(f"WRN: {wav_name} was not generated. Skipping this track.")
            
    video["Sound"] = new_sounds
    
    # 2. Render final video
    video_name = f"{video['CgName']}_{video['GirlOrBoy']}_{get_filename_no_ext(video['CgFile'])}.mp4"
    out_path = VIDEOS_DIR / video_name
    
    if out_path.exists():
        if progress_callback: progress_callback(1)
        return True
        
    video_file = video["CgFile"]
    audio_files = video["Sound"]
    audio_count = len(audio_files)
    
    srt_path = CAPTIONS_DIR / f"{video['CgName']}.srt"
    has_srt = (subtitle_mode != "None") and srt_path.exists()
    is_hardsub = has_srt and (subtitle_mode == "Hard-sub")
    is_softsub = has_srt and (subtitle_mode == "Soft-sub")
    
    ffmpeg_exe = get_ffmpeg_path()
    cmd = [ffmpeg_exe, "-y", "-i", video_file]
    
    if audio_count == 0:
        if is_softsub:
            cmd += ["-i", str(srt_path), "-map", "0:v:0", "-map", "1:s:0", "-c", "copy", "-c:s", "mov_text", str(out_path)]
        elif is_hardsub:
            # Escape path for video filter: replace backward slashes with forward slashes and escape colon
            escaped_srt = str(srt_path).replace('\\', '/').replace(':', '\\:')
            cmd += ["-vf", f"subtitles='{escaped_srt}':force_style='Fontname=Kanit Medium,Outline=1,Shadow=0'", "-map", "0:v:0", "-c:v", "libx264", str(out_path)]
        else:
            cmd += ["-c", "copy", str(out_path)]
            
    elif audio_count == 1:
        cmd += ["-i", audio_files[0]]
        if is_softsub: cmd += ["-i", str(srt_path)]
        
        cmd += ["-filter_complex", "aformat=channel_layouts=stereo[aout]", "-map", "0:v:0", "-map", "[aout]"]
        
        if is_softsub: 
            cmd += ["-map", "2:s:0", "-c:s", "mov_text"]
            
        if is_hardsub:
            escaped_srt = str(srt_path).replace('\\', '/').replace(':', '\\:')
            cmd += ["-vf", f"subtitles='{escaped_srt}':force_style='Fontname=Kanit Medium,Outline=1,Shadow=0'", "-c:v", "libx264", "-c:a", "aac"]
        else:
            cmd += ["-c:v", "copy", "-c:a", "aac"]
            
        if not has_srt: cmd += ["-shortest"]
        cmd += [str(out_path)]
        
    else:
        for a in audio_files:
            cmd += ["-i", a]
        if is_softsub:
            cmd += ["-i", str(srt_path)]
            srt_index = 1 + audio_count
            
        amix_inputs = "".join(f"[{j + 1}:a]" for j in range(audio_count))
        filter_complex = f"{amix_inputs}amix=inputs={audio_count}:duration=longest,aresample=async=1,aformat=channel_layouts=stereo[aout]"
        
        if is_hardsub:
            escaped_srt = str(srt_path).replace('\\', '/').replace(':', '\\:')
            filter_complex = f"subtitles='{escaped_srt}':force_style='Fontname=Kanit Medium,Outline=2,Shadow=0';{filter_complex}"
            
        cmd += ["-filter_complex", filter_complex, "-map", "0:v:0", "-map", "[aout]"]
        
        if is_softsub: 
            cmd += ["-map", f"{srt_index}:s:0", "-c:s", "mov_text"]
            
        if is_hardsub:
            cmd += ["-c:v", "libx264", "-c:a", "aac"]
        else:
            cmd += ["-c:v", "copy", "-c:a", "aac"]
            
        if not has_srt: cmd += ["-shortest"]
        cmd += [str(out_path)]
        
    if log_callback: log_callback(f"Rendering {video_name}...")
    
    process = subprocess.Popen(cmd, stdout=subprocess.DEVNULL, stderr=subprocess.PIPE, text=True, bufsize=1, universal_newlines=True)
    error_lines = []
    for line in iter(process.stderr.readline, ''):
        if line:
            if not ("time=" in line or "frame=" in line):
                error_lines.append(line.strip('\r\n'))
                
    return_code = process.wait()
    
    if return_code != 0:
        if log_callback: 
            log_callback(f"ERR: ffmpeg failed for {video_name}")
            for err in error_lines[-5:]:
                log_callback(f"ffmpeg err: {err}")
    else:
        if log_callback: log_callback(f"Finished {video_name}")
                
    if progress_callback: progress_callback(1)
    return True

def render_videos(videos_info, subtitle_mode="Soft-sub", custom_bgm="", log_callback=None, progress_callback=None):
    if log_callback: log_callback(f"Starting Multi-threaded Render (Mode: {subtitle_mode})...")
    
    total = len(videos_info)
    completed = 0
    lock = Lock()
    
    def locked_progress(step):
        nonlocal completed
        with lock:
            completed += step
            if progress_callback: progress_callback(completed / total)

    with concurrent.futures.ThreadPoolExecutor(max_workers=3) as executor:
        futures = []
        for i, video in enumerate(videos_info):
            # Inject custom BGM if provided
            if custom_bgm and os.path.exists(custom_bgm):
                video["Sound"].append(custom_bgm)
                
            futures.append(executor.submit(render_single_video, i, video, total, subtitle_mode, log_callback, locked_progress))
        
        for future in concurrent.futures.as_completed(futures):
            future.result()
            
    # Clean up temporary .wav files
    if log_callback: log_callback("Cleaning up temporary audio files...")
    for f in os.listdir(VIDEOS_DIR):
        if f.endswith('.wav'):
            try:
                os.remove(VIDEOS_DIR / f)
            except Exception as e:
                pass
                
    if log_callback: log_callback("All videos rendered successfully!")
    return True

File: tauri-app/src-tauri/src/commands/renderer.rs
use std::fs;
use std::path::{PathBuf};
use std::process::Command;
use tauri::AppHandle;

use crate::commands::downloader::get_tools_dir;
use crate::commands::generator::VideoInfo;
use crate::commands::log::emit_log;

#[tauri::command]
pub async fn render_video(
    app: AppHandle,
    video_info: VideoInfo,
    subtitle_mode: String,
) -> Result<(), String> {
    let tools_dir = get_tools_dir(&app);
    let vgmstream_exe = tools_dir.join("vgmstream-cli.exe");
    let ffmpeg_exe = tools_dir.join("ffmpeg.exe");

    if !vgmstream_exe.exists() { return Err("vgmstream-cli.exe not found".into()); }
    if !ffmpeg_exe.exists() { return Err("ffmpeg.exe not found".into()); }

    let output_dir = crate::commands::downloader::get_data_dir(&app).join("output");
    let _ = fs::create_dir_all(&output_dir);

    // 1. Convert txtp to wav
    let mut wav_paths = Vec::new();
    for sound in &video_info.sounds {
        let sound_path = PathBuf::from(sound);
        let wav_name = format!("{}.wav", sound_path.file_stem().unwrap().to_string_lossy());
        let wav_path = output_dir.join(&wav_name);
        
        if !wav_path.exists() {
            emit_log(&app, &format!("Converting {} to wav...", sound_path.file_name().unwrap().to_string_lossy()));
            let status = Command::new(&vgmstream_exe)
                .args(["-o", &wav_path.to_string_lossy(), &sound_path.to_string_lossy()])
                .status()
                .map_err(|e| e.to_string())?;
                
            if !status.success() {
                emit_log(&app, &format!("Warning: vgmstream failed for {}", sound));
            }
        }
        
        if wav_path.exists() {
            wav_paths.push(wav_path);
        }
    }

    // 2. FFmpeg Render
    let video_path = PathBuf::from(&video_info.cg_file);
    let out_name = format!("{}_{}_{}.mp4", video_info.cg_name, video_info.girl_or_boy, video_path.file_stem().unwrap().to_string_lossy());
    let out_path = output_dir.join(&out_name);

    if out_path.exists() {
        emit_log(&app, &format!("Video {} already exists. Skipping.", out_name));
        return Ok(());
    }

    emit_log(&app, &format!("Rendering {}...", out_name));

    let mut cmd = Command::new(&ffmpeg_exe);
    cmd.args(["-y", "-i", &video_info.cg_file]);

    let audio_count = wav_paths.len();
    for wav in &wav_paths {
        cmd.args(["-i", &wav.to_string_lossy()]);
    }

    let has_srt = subtitle_mode != "None" && video_info.srt_path.is_some();
    let is_hardsub = has_srt && subtitle_mode == "Hard-sub";
    let is_softsub = has_srt && subtitle_mode == "Soft-sub";
    
    let srt_path = video_info.srt_path.unwrap_or_default();
    let mut srt_index = 1 + audio_count;

    if is_softsub {
        cmd.args(["-i", &srt_path]);
    }

    if audio_count == 0 {
        if is_softsub {
            cmd.args(["-map", "0:v:0", "-map", "1:s:0", "-c", "copy", "-c:s", "mov_text"]);
        } else if is_hardsub {
            let escaped_srt = srt_path.replace("\\", "/").replace(":", "\\:");
            cmd.args(["-vf", &format!("subtitles='{}':force_style='Fontname=Kanit Medium,Outline=1,Shadow=0'", escaped_srt), "-map", "0:v:0", "-c:v", "libx264"]);
        } else {
            cmd.args(["-c", "copy"]);
        }
    } else if audio_count == 1 {
        cmd.args(["-filter_complex", "aformat=channel_layouts=stereo[aout]", "-map", "0:v:0", "-map", "[aout]"]);
        
        if is_softsub {
            cmd.args(["-map", &format!("{}:s:0", srt_index), "-c:s", "mov_text"]);
        }
        if is_hardsub {
            let escaped_srt = srt_path.replace("\\", "/").replace(":", "\\:");
            cmd.args(["-vf", &format!("subtitles='{}':force_style='Fontname=Kanit Medium,Outline=1,Shadow=0'", escaped_srt), "-c:v", "libx264", "-c:a", "aac"]);
        } else {
            cmd.args(["-c:v", "copy", "-c:a", "aac"]);
        }
        if !has_srt { cmd.arg("-shortest"); }
    } else {
        let mut amix_inputs = String::new();
        for j in 0..audio_count {
            amix_inputs.push_str(&format!("[{}:a]", j + 1));
        }
        let mut filter_complex = format!("{}amix=inputs={}:duration=longest,aresample=async=1,aformat=channel_layouts=stereo[aout]", amix_inputs, audio_count);
        
        if is_hardsub {
            let escaped_srt = srt_path.replace("\\", "/").replace(":", "\\:");
            filter_complex = format!("subtitles='{}':force_style='Fontname=Kanit Medium,Outline=2,Shadow=0';{}", escaped_srt, filter_complex);
        }
        
        cmd.args(["-filter_complex", &filter_complex, "-map", "0:v:0", "-map", "[aout]"]);
        
        if is_softsub {
            cmd.args(["-map", &format!("{}:s:0", srt_index), "-c:s", "mov_text"]);
        }
        if is_hardsub {
            cmd.args(["-c:v", "libx264", "-c:a", "aac"]);
        } else {
            cmd.args(["-c:v", "copy", "-c:a", "aac"]);
        }
        if !has_srt { cmd.arg("-shortest"); }
    }

    cmd.arg(&out_path);

    let output = cmd.output().map_err(|e| e.to_string())?;
    
    if !output.status.success() {
        let stderr = String::from_utf8_lossy(&output.stderr);
        emit_log(&app, &format!("FFmpeg error: {}", stderr));
        return Err("Video rendering failed".into());
    }

    emit_log(&app, &format!("Finished {}", out_name));

    // Cleanup temporary wav files
    for wav in wav_paths {
        let _ = fs::remove_file(wav);
    }

    Ok(())
}
